## Tasca M3 T02 Programació Numèrica amb Pandas

**Descripció:** Exercicis amb Dataframes.


- **Exercici 1**

    Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

    ->Airlines Delay: Airline on-time statistics and delay causes

    **Important**

    Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.


- **Exercici 2**

    Fes un informe complet del dataset:

    Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
    Troba quantes dades faltants hi ha per columna.
    Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
    Fes una taula de les aerolínies amb més endarreriments acumulats.
    Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
    Aporta allò que consideris rellevant.


- **Exercici 3**

    Exporta el dataset net i amb les noves columnes a Excel.

In [164]:
##µImporting libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import numexpr

## Exercici 1
<div class="alert alert-block alert-info">Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.    
</div>

**Important**: Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

## Flights Delay DataSet description:

This dataset is composed by the following variables:
From :https://www.kaggle.com/code/adveros/flight-delay-eda-exploratory-data-analysis

- Year 2008
- Month 1-12
- DayofMonth 1-31
- DayOfWeek 1 (Monday) - 7 (Sunday)
- DepTime actual departure time (local, hhmm)
- CRSDepTime scheduled departure time (local, hhmm)
- ArrTime actual arrival time (local, hhmm)
- CRSArrTime scheduled arrival time (local, hhmm)
- UniqueCarrier unique carrier code
- FlightNum flight number
- TailNum plane tail number: aircraft registration, unique aircraft identifier
- ActualElapsedTime in minutes
- CRSElapsedTime in minutes
- AirTime in minutes
- ArrDelay arrival delay, in minutes: A flight is counted as "on time" if it operated less than 15 minutes later the scheduled time shown in the carriers' Computerized Reservations Systems (CRS).
- DepDelay departure delay, in minutes
- Origin origin IATA airport code
- Dest destination IATA airport code
- Distance in miles
- TaxiIn taxi in time, in minutes
- TaxiOut taxi out time in minutes
- Cancelled *was the flight cancelled
- CancellationCode reason for cancellation (A = carrier, B = weather, C = NAS, D = security)
- Diverted 1 = yes, 0 = no
- CarrierDelay in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.
- WeatherDelay in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.
- NASDelay in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.
- SecurityDelay in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.
- LateAircraftDelay in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.



In [8]:
df = pd.read_csv('DelayedFlights.csv') #read data from CSV file
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [51]:
##I take a sample of the dataset of 200000 registers
df = df.sample(n=200000)

In [52]:
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
594306      1857022  2008      4          23          3   1555.0        1515   
1878764     6849172  2008     12          21          7   1612.0        1435   
86433        274177  2008      1          15          2   2250.0        2215   
1524919     5388116  2008     10           4          6   1015.0        1005   
1049925     3508030  2008      6           6          5   1915.0        1725   
...             ...   ...    ...         ...        ...      ...         ...   
1643162     6017836  2008     11          26          3   1114.0        1105   
452642      1412905  2008      3          22          6   1222.0        1215   
368057      1154269  2008      2          27          3   1821.0        1803   
1893559     6894075  2008     12          15          1   1007.0         937   
324998      1027633  2008      2           4          1   2232.0        2225   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
594306    1658.0        1605            WN  ...     3.0    25.0          0   
1878764   1719.0        1559            MQ  ...     7.0    19.0          0   
86433       48.0          24            US  ...     6.0    13.0          0   
1524919   1106.0        1100            WN  ...     2.0     6.0          0   
1049925   2040.0        1850            AA  ...     3.0    16.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1643162   1404.0        1355            WN  ...     6.0     7.0          0   
452642    1521.0        1507            UA  ...     5.0    29.0          0   
368057    1935.0        1953            CO  ...     6.0    14.0          0   
1893559   1116.0        1000            9E  ...     5.0    60.0          0   
324998    2358.0        2357            NW  ...     6.0    10.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
594306                  N         0           0.0           0.0     13.0   
1878764                 N         0           0.0           0.0     49.0   
86433                   N         0           0.0           0.0      0.0   
1524919                 N         0           NaN           NaN      NaN   
1049925                 N         0           0.0           0.0     46.0   
...                   ...       ...           ...           ...      ...   
1643162                 N         0           NaN           NaN      NaN   
452642                  N         0           NaN           NaN      NaN   
368057                  N         0           NaN           NaN      NaN   
1893559                 N         0          30.0           0.0     46.0   
324998                  N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
594306            0.0               40.0  
1878764           0.0               31.0  
86433             0.0               24.0  
1524919           NaN                NaN  
1049925           0.0               64.0  
...               ...                ...  
1643162           NaN                NaN  
452642            NaN                NaN  
368057            NaN                NaN  
1893559           0.0                0.0  
324998            NaN                NaN  

[200000 rows x 30 columns]

In [53]:
df.columns ##Check df columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [54]:
df.info() ##Check for type of variables in each column

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 594306 to 324998
Data columns (total 30 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         200000 non-null  int64  
 1   Year               200000 non-null  int64  
 2   Month              200000 non-null  int64  
 3   DayofMonth         200000 non-null  int64  
 4   DayOfWeek          200000 non-null  int64  
 5   DepTime            200000 non-null  float64
 6   CRSDepTime         200000 non-null  int64  
 7   ArrTime            199297 non-null  float64
 8   CRSArrTime         200000 non-null  int64  
 9   UniqueCarrier      200000 non-null  object 
 10  FlightNum          200000 non-null  int64  
 11  TailNum            199999 non-null  object 
 12  ActualElapsedTime  199169 non-null  float64
 13  CRSElapsedTime     199984 non-null  float64
 14  AirTime            199169 non-null  float64
 15  ArrDelay           199169 non-null  float64
 1

In [55]:
df.describe() ##Get a statistical basic description for numerical variables

Unnamed: 0      Year          Month     DayofMonth     DayOfWeek  \
count  2.000000e+05  200000.0  200000.000000  200000.000000  200000.00000   
mean   3.342493e+06    2008.0       6.113375      15.757380       3.98304   
std    2.067080e+06       0.0       3.483511       8.770968       1.99635   
min    4.300000e+01    2008.0       1.000000       1.000000       1.00000   
25%    1.511407e+06    2008.0       3.000000       8.000000       2.00000   
50%    3.247595e+06    2008.0       6.000000      16.000000       4.00000   
75%    4.971760e+06    2008.0       9.000000      23.000000       6.00000   
max    7.009717e+06    2008.0      12.000000      31.000000       7.00000   

             DepTime     CRSDepTime        ArrTime     CRSArrTime  \
count  200000.000000  200000.000000  199297.000000  200000.000000   
mean     1519.044285    1468.508485    1609.078350    1634.845315   
std       450.928465     424.665441     549.671022     464.862733   
min         1.000000       1.000000       1.000000       0.000000   
25%      1203.000000    1135.000000    1315.000000    1325.000000   
50%      1545.000000    1510.000000    1715.000000    1707.000000   
75%      1901.000000    1820.000000    2031.000000    2015.000000   
max      2400.000000    2359.000000    2400.000000    2400.000000   

           FlightNum  ...       Distance         TaxiIn        TaxiOut  \
count  200000.000000  ...  200000.000000  199297.000000  199945.000000   
mean     2186.217045  ...     766.431290       6.828332      18.191838   
std      1945.166121  ...     574.376858       5.400517      14.166003   
min         1.000000  ...      31.000000       0.000000       0.000000   
25%       613.000000  ...     341.000000       4.000000      10.000000   
50%      1545.000000  ...     607.000000       6.000000      14.000000   
75%      3423.000000  ...     998.000000       8.000000      21.000000   
max      9741.000000  ...    4962.000000     225.000000     321.000000   

           Cancelled       Diverted   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.000000  128859.000000  128859.000000   
mean        0.000345       0.003810      18.938537       3.671013   
std         0.018571       0.061608      42.805013      20.999252   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       2.000000       0.000000   
75%         0.000000       0.000000      21.000000       0.000000   
max         1.000000       1.000000    1510.000000     940.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  128859.000000  128859.000000      128859.000000  
mean       14.959514       0.088290          25.539861  
std        33.346709       1.826137          42.410760  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         2.000000       0.000000           8.000000  
75%        15.000000       0.000000          34.000000  
max       621.000000     211.000000        1143.000000  

[8 rows x 25 columns]

I use the round methode with 2 decimals in order to have more readable table 
from: https://datatofish.com/round-values-pandas-dataframe/
- Round methode could be a problem with some statistical values (standard deviation), but it doesn't look like here.
- I could also use: 
    - 2) Round up values under a single DataFrame column. df['DataFrame column'].apply(np.ceil)
    - 3) Round down values under a single DataFrame column. df['DataFrame column'].apply(np.floor)

In [56]:
df.describe().round(2) 

Unnamed: 0      Year      Month  DayofMonth  DayOfWeek    DepTime  \
count   200000.00  200000.0  200000.00   200000.00  200000.00  200000.00   
mean   3342493.11    2008.0       6.11       15.76       3.98    1519.04   
std    2067080.06       0.0       3.48        8.77       2.00     450.93   
min         43.00    2008.0       1.00        1.00       1.00       1.00   
25%    1511407.00    2008.0       3.00        8.00       2.00    1203.00   
50%    3247595.00    2008.0       6.00       16.00       4.00    1545.00   
75%    4971760.00    2008.0       9.00       23.00       6.00    1901.00   
max    7009717.00    2008.0      12.00       31.00       7.00    2400.00   

       CRSDepTime    ArrTime  CRSArrTime  FlightNum  ...   Distance  \
count   200000.00  199297.00   200000.00  200000.00  ...  200000.00   
mean      1468.51    1609.08     1634.85    2186.22  ...     766.43   
std        424.67     549.67      464.86    1945.17  ...     574.38   
min          1.00       1.00        0.00       1.00  ...      31.00   
25%       1135.00    1315.00     1325.00     613.00  ...     341.00   
50%       1510.00    1715.00     1707.00    1545.00  ...     607.00   
75%       1820.00    2031.00     2015.00    3423.00  ...     998.00   
max       2359.00    2400.00     2400.00    9741.00  ...    4962.00   

          TaxiIn    TaxiOut  Cancelled   Diverted  CarrierDelay  WeatherDelay  \
count  199297.00  199945.00  200000.00  200000.00     128859.00     128859.00   
mean        6.83      18.19       0.00       0.00         18.94          3.67   
std         5.40      14.17       0.02       0.06         42.81         21.00   
min         0.00       0.00       0.00       0.00          0.00          0.00   
25%         4.00      10.00       0.00       0.00          0.00          0.00   
50%         6.00      14.00       0.00       0.00          2.00          0.00   
75%         8.00      21.00       0.00       0.00         21.00          0.00   
max       225.00     321.00       1.00       1.00       1510.00        940.00   

        NASDelay  SecurityDelay  LateAircraftDelay  
count  128859.00      128859.00          128859.00  
mean       14.96           0.09              25.54  
std        33.35           1.83              42.41  
min         0.00           0.00               0.00  
25%         0.00           0.00               0.00  
50%         2.00           0.00               8.00  
75%        15.00           0.00              34.00  
max       621.00         211.00            1143.00  

[8 rows x 25 columns]

In [57]:
df.shape

(200000, 30)

In [58]:
df.isnull().sum() #Checking how many null rowws we have in each column

Unnamed: 0               0
Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                703
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  1
ActualElapsedTime      831
CRSElapsedTime          16
AirTime                831
ArrDelay               831
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 703
TaxiOut                 55
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71141
WeatherDelay         71141
NASDelay             71141
SecurityDelay        71141
LateAircraftDelay    71141
dtype: int64

**&rarr; Now I will check categorical variables**

In [59]:
df.UniqueCarrier.unique() ##Checking the unic values in UniqueCarrier

array(['WN', 'MQ', 'US', 'AA', 'EV', 'XE', 'DL', 'UA', 'AS', 'OO', 'FL',
       'OH', 'NW', 'CO', '9E', 'HA', 'YV', 'B6', 'F9', 'AQ'], dtype=object)

In [60]:
df.UniqueCarrier.value_counts() ## Checking how many times appear each of the unique values

WN    39164
AA    19821
UA    14646
MQ    14569
OO    13617
DL    11896
XE    10732
CO    10232
US    10194
EV     8430
NW     8197
FL     7488
YV     6931
B6     5571
OH     5515
9E     5275
AS     4038
F9     2839
HA      772
AQ       73
Name: UniqueCarrier, dtype: int64

In [61]:
df.TailNum.unique()

array(['N524SW', 'N851MQ', 'N423US', ..., 'N797UA', 'N817NW', 'N180DN'],
      dtype=object)

In [62]:
df.TailNum.value_counts()

N768AS    115
N612SW    112
N676SW    112
N347SW    111
N663SW    109
         ... 
N866DA      1
N814NW      1
N121UA      1
N801MQ      1
N180DN      1
Name: TailNum, Length: 5266, dtype: int64

In [63]:
df.Origin.unique()

array(['DAL', 'DCA', 'CLT', 'BWI', 'ORD', 'MCI', 'ATL', 'SAN', 'MHT',
       'PHX', 'DFW', 'BUF', 'SMF', 'DEN', 'OMA', 'LGA', 'FLL', 'GRR',
       'ABQ', 'ONT', 'SEA', 'MKE', 'PSP', 'STL', 'MCO', 'IAH', 'LAX',
       'MIA', 'HOU', 'MSN', 'EWR', 'FWA', 'SDF', 'TUS', 'MDW', 'XNA',
       'PHL', 'HNL', 'IND', 'LAS', 'FAT', 'TVC', 'MEM', 'TPA', 'DTW',
       'CVG', 'BOI', 'MSP', 'CMH', 'ORF', 'JAN', 'JFK', 'JAX', 'BNA',
       'MSY', 'SLC', 'FSD', 'SFO', 'BDL', 'CRW', 'PVD', 'BOS', 'RSW',
       'ROW', 'OAK', 'COS', 'CAK', 'IAD', 'TOL', 'TYS', 'DAY', 'PBI',
       'PIT', 'SAT', 'CMI', 'BET', 'RNO', 'LBB', 'HPN', 'LIH', 'CAE',
       'CLE', 'TLH', 'RDU', 'FCA', 'SYR', 'MBS', 'GRK', 'SNA', 'OKC',
       'AUS', 'PDX', 'MGM', 'SJC', 'LAN', 'KOA', 'ANC', 'CHS', 'MLI',
       'GEG', 'ELP', 'TUL', 'SAV', 'ISP', 'BHM', 'JAC', 'HRL', 'LGB',
       'ROC', 'SBA', 'RDD', 'SJU', 'RIC', 'PSC', 'CHA', 'MTJ', 'ICT',
       'BUR', 'GUC', 'ALB', 'MAF', 'EVV', 'BTV', 'ABE', 'LIT', 'SPI',
       'AGS', 'GSP',

In [64]:
df.Origin.value_counts()

ATL    13632
ORD    13012
DFW     9812
DEN     7517
LAX     6098
       ...  
HTS        2
GST        2
ACY        2
SUX        1
BJI        1
Name: Origin, Length: 298, dtype: int64

In [65]:
df.Dest.unique()

array(['AUS', 'JFK', 'MHT', 'PIT', 'MCI', 'PDX', 'DAY', 'ALB', 'LGA',
       'SMF', 'EWR', 'LAS', 'HNL', 'MDW', 'DEN', 'OKC', 'SNA', 'SFO',
       'SEA', 'ORD', 'MKE', 'DFW', 'ATL', 'SJU', 'SJC', 'CLT', 'CLE',
       'MSP', 'PHL', 'PHX', 'MIA', 'MSY', 'DCA', 'OMA', 'BNA', 'LIT',
       'RIC', 'MDT', 'BOS', 'SLC', 'SAN', 'LAX', 'DTW', 'BWI', 'MCO',
       'KOA', 'SHV', 'BDL', 'GSO', 'VPS', 'DAL', 'FAT', 'SAT', 'GRB',
       'IAH', 'TPA', 'FSD', 'RDU', 'BHM', 'FLL', 'BUF', 'GSP', 'PVD',
       'LSE', 'ATW', 'IND', 'FAY', 'ORF', 'PSP', 'CHS', 'STL', 'AZO',
       'ONT', 'COS', 'HDN', 'HOU', 'TUL', 'BUR', 'OAK', 'ABQ', 'SBA',
       'GPT', 'DSM', 'BQN', 'IAD', 'CAK', 'ANC', 'RNO', 'FSM', 'JAN',
       'RSW', 'ISP', 'PBI', 'BMI', 'MYR', 'MSN', 'BOI', 'ICT', 'LAN',
       'CVG', 'CMH', 'SRQ', 'PIA', 'EUG', 'IDA', 'TLH', 'MEM', 'PHF',
       'LFT', 'LIH', 'JAX', 'OTZ', 'MOB', 'CRP', 'ACK', 'SWF', 'SAV',
       'SDF', 'GRK', 'GRR', 'LGB', 'EKO', 'STT', 'HPN', 'CID', 'OGG',
       'BTR', 'LRD',

In [66]:
df.Dest.value_counts()

ORD    11344
ATL    11241
DFW     7198
DEN     6568
LAX     6179
       ...  
BJI        1
INL        1
GST        1
TUP        1
TEX        1
Name: Dest, Length: 300, dtype: int64

In [67]:
df.CancellationCode.unique()

array(['N', 'A', 'C', 'B'], dtype=object)

In [68]:
df.CancellationCode.value_counts()

N    199931
B        30
A        28
C        11
Name: CancellationCode, dtype: int64

In [69]:
df.Cancelled.unique()

array([0, 1], dtype=int64)

In [114]:
df.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

## Preliminar data analysis

After checking the main basic columns, tpe of data and numerical variables basic statistics, and existence of null values, we proceed to concider which columns are usefull and whcih not.


- First, we have an **unamed column** that doesn't look relevant and it is not an index number so, I will drop it
- Then we have the **year, month, dayofmonth, dayofweek**, with no nulls and referring to the flight timing.
- Then, We have information on **departure and arrival times**. 
    - **Departure times and arrival** times go from 0 to 24, which I interprete hours of the day.
    - Whe see there are 7003 nulls arrival times, the same numbers as Tax in columns,  which probably correspond to flights that have been delayed for the whole 24h due to external reasons. 
    - **CRS** arrival and departure are the Scheduled times, important to understand the departure and arrival delays or earliness/promptness
- Then we have information of the flight itself, the **Carrier**, and the **flight and tail numbers**. 
    - Unique carrier looks important to check which carrier got more delays, problemes..etc
    - But flight and tail numbers (the tail is not a number is an object...) doesn't look relevant. I will drop the tail number and see if flight number is relevant
- Whe have 3 columns with time of flight original and real **ActualElapsedTime, CRSElapsedTime, AirTime**. Which I dont thing will be useful for a basinc data analysis. For more andvanced probably will be. Today I will drop them.
- We already have calculated the delays With **arrDelay and DepDelay**, so we could probably drop all arrival and departure times, but here we have several null in most of them, so maybe I could recalculated them and see where the nulls come from. 
- Then **Origine, Destination and Distance** information, the first two are categorical variable, and destination is an integer. This can be useful to analyze in which airports you can get more delays because of weather, taxi in and out, or maybe distance of the flight?
- **Taxi-in, Taxi-out** is the time the plane stays in airport between flights (jumps) This will be relevant to compare the delay between flights (jumps) which I am not going todo today...So probably drop them.We also have the same number of null values in Taxi in as arrival time, so agin this could be more than 24h delay due to external reasons.

- Then we have if the flight was **Cancelled** with integer 0, 1 options and no nulls. And a **CancellationCode** categorical with 4 options quite criptic, so probably I'll drop them. And **Diverted** flights which at the moment I don't find relevant for my analysis.

- Finally, we get 5 columns **'CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'** related to the type of delay all of them have the same number of nulls (71141). I understand this because if there is delay, it is only for one reason, so they are exclusive. 


I proceed with droping the non ionteresting columns

In [269]:
##return a new DataFrame with the desired columns of df so I keep the original intact
flights_df = df[['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
                'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
                'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
                'DepDelay', 'Origin', 'Dest', 'Distance', 'Cancelled', 'CarrierDelay','WeatherDelay', 
                'NASDelay', 'SecurityDelay', 'LateAircraftDelay']] 
flights_df 

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
594306   2008      4          23          3   1555.0        1515   1658.0   
1878764  2008     12          21          7   1612.0        1435   1719.0   
86433    2008      1          15          2   2250.0        2215     48.0   
1524919  2008     10           4          6   1015.0        1005   1106.0   
1049925  2008      6           6          5   1915.0        1725   2040.0   
...       ...    ...         ...        ...      ...         ...      ...   
1643162  2008     11          26          3   1114.0        1105   1404.0   
452642   2008      3          22          6   1222.0        1215   1521.0   
368057   2008      2          27          3   1821.0        1803   1935.0   
1893559  2008     12          15          1   1007.0         937   1116.0   
324998   2008      2           4          1   2232.0        2225   2358.0   

         CRSArrTime UniqueCarrier  FlightNum  ...  DepDelay  Origin  Dest  \
594306         1605            WN        649  ...      40.0     DAL   AUS   
1878764        1559            MQ       4750  ...      97.0     DCA   JFK   
86433            24            US       1116  ...      35.0     CLT   MHT   
1524919        1100            WN        185  ...      10.0     BWI   PIT   
1049925        1850            AA       1618  ...     110.0     ORD   MCI   
...             ...           ...        ...  ...       ...     ...   ...   
1643162        1355            WN       3116  ...       9.0     OAK   PHX   
452642         1507            UA        490  ...       7.0     DEN   DFW   
368057         1953            CO       1600  ...      18.0     PHL   CLE   
1893559        1000            9E       5821  ...      30.0     ATL   JAN   
324998         2357            NW       1469  ...       7.0     MSP   MOT   

         Distance  Cancelled CarrierDelay WeatherDelay  NASDelay  \
594306        189          0          0.0          0.0      13.0   
1878764       213          0          0.0          0.0      49.0   
86433         737          0          0.0          0.0       0.0   
1524919       210          0          NaN          NaN       NaN   
1049925       403          0          0.0          0.0      46.0   
...           ...        ...          ...          ...       ...   
1643162       646          0          NaN          NaN       NaN   
452642        641          0          NaN          NaN       NaN   
368057        363          0          NaN          NaN       NaN   
1893559       341          0         30.0          0.0      46.0   
324998        449          0          NaN          NaN       NaN   

         SecurityDelay  LateAircraftDelay  
594306             0.0               40.0  
1878764            0.0               31.0  
86433              0.0               24.0  
1524919            NaN                NaN  
1049925            0.0               64.0  
...                ...                ...  
1643162            NaN                NaN  
452642             NaN                NaN  
368057             NaN                NaN  
1893559            0.0                0.0  
324998             NaN                NaN  

[200000 rows x 24 columns]

I don't like the index column because it kept the  number it had before the sampling of the whole data set.
I will resett the index column and put it in place.
But I'm not entirely sure this is the best option, maibe I should keep the old one and assigne a new one...?¿

In [270]:
flights_df.reset_index(drop=True, inplace=True) ##Resetting of index column, droping old true, in the same place true
flights_df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0       2008      4          23          3   1555.0        1515   1658.0   
1       2008     12          21          7   1612.0        1435   1719.0   
2       2008      1          15          2   2250.0        2215     48.0   
3       2008     10           4          6   1015.0        1005   1106.0   
4       2008      6           6          5   1915.0        1725   2040.0   
...      ...    ...         ...        ...      ...         ...      ...   
199995  2008     11          26          3   1114.0        1105   1404.0   
199996  2008      3          22          6   1222.0        1215   1521.0   
199997  2008      2          27          3   1821.0        1803   1935.0   
199998  2008     12          15          1   1007.0         937   1116.0   
199999  2008      2           4          1   2232.0        2225   2358.0   

        CRSArrTime UniqueCarrier  FlightNum  ...  DepDelay  Origin  Dest  \
0             1605            WN        649  ...      40.0     DAL   AUS   
1             1559            MQ       4750  ...      97.0     DCA   JFK   
2               24            US       1116  ...      35.0     CLT   MHT   
3             1100            WN        185  ...      10.0     BWI   PIT   
4             1850            AA       1618  ...     110.0     ORD   MCI   
...            ...           ...        ...  ...       ...     ...   ...   
199995        1355            WN       3116  ...       9.0     OAK   PHX   
199996        1507            UA        490  ...       7.0     DEN   DFW   
199997        1953            CO       1600  ...      18.0     PHL   CLE   
199998        1000            9E       5821  ...      30.0     ATL   JAN   
199999        2357            NW       1469  ...       7.0     MSP   MOT   

        Distance  Cancelled CarrierDelay WeatherDelay  NASDelay  \
0            189          0          0.0          0.0      13.0   
1            213          0          0.0          0.0      49.0   
2            737          0          0.0          0.0       0.0   
3            210          0          NaN          NaN       NaN   
4            403          0          0.0          0.0      46.0   
...          ...        ...          ...          ...       ...   
199995       646          0          NaN          NaN       NaN   
199996       641          0          NaN          NaN       NaN   
199997       363          0          NaN          NaN       NaN   
199998       341          0         30.0          0.0      46.0   
199999       449          0          NaN          NaN       NaN   

        SecurityDelay  LateAircraftDelay  
0                 0.0               40.0  
1                 0.0               31.0  
2                 0.0               24.0  
3                 NaN                NaN  
4                 0.0               64.0  
...               ...                ...  
199995            NaN                NaN  
199996            NaN                NaN  
199997            NaN                NaN  
199998            0.0                0.0  
199999            NaN                NaN  

[200000 rows x 24 columns]

## Exercici 2
<div class="alert alert-block alert-info"> Fes un informe complet del dataset:

- Resumeix estadísticament el dataset i les columnes d’interès. 
    
- Fes una anàlisi estadístic del que consideris rellevant. 

- Troba quantes dades faltants hi ha per columna. 

- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...). 
    
- Fes una taula de les aerolínies amb més endarreriments acumulats. 
    
- Quins són els vols més llargs? I els més endarrerits? 
    
- Busca les rutes més llargues i les que acumulen més retards. 
    
- Aporta allò que consideris rellevant.    
</div>

**&rarr; Part of the exercice was done before, so I will just repeat some basic statistical analysis with describe and count of nulls**

In [271]:
flights_df.describe().round(2) ##for basic statistical description of numerical variables

Year      Month  DayofMonth  DayOfWeek    DepTime  CRSDepTime  \
count  200000.0  200000.00   200000.00  200000.00  200000.00   200000.00   
mean     2008.0       6.11       15.76       3.98    1519.04     1468.51   
std         0.0       3.48        8.77       2.00     450.93      424.67   
min      2008.0       1.00        1.00       1.00       1.00        1.00   
25%      2008.0       3.00        8.00       2.00    1203.00     1135.00   
50%      2008.0       6.00       16.00       4.00    1545.00     1510.00   
75%      2008.0       9.00       23.00       6.00    1901.00     1820.00   
max      2008.0      12.00       31.00       7.00    2400.00     2359.00   

         ArrTime  CRSArrTime  FlightNum  ActualElapsedTime  ...    AirTime  \
count  199297.00   200000.00  200000.00          199169.00  ...  199169.00   
mean     1609.08     1634.85    2186.22             133.37  ...     108.37   
std       549.67      464.86    1945.17              72.01  ...      68.60   
min         1.00        0.00       1.00              16.00  ...       1.00   
25%      1315.00     1325.00     613.00              80.00  ...      58.00   
50%      1715.00     1707.00    1545.00             116.00  ...      90.00   
75%      2031.00     2015.00    3423.00             165.00  ...     137.00   
max      2400.00     2400.00    9741.00             681.00  ...     654.00   

        ArrDelay   DepDelay   Distance  Cancelled  CarrierDelay  WeatherDelay  \
count  199169.00  200000.00  200000.00  200000.00     128859.00     128859.00   
mean       42.14      43.14     766.43       0.00         18.94          3.67   
std        56.21      52.81     574.38       0.02         42.81         21.00   
min       -70.00       6.00      31.00       0.00          0.00          0.00   
25%         9.00      12.00     341.00       0.00          0.00          0.00   
50%        24.00      24.00     607.00       0.00          2.00          0.00   
75%        56.00      53.00     998.00       0.00         21.00          0.00   
max      1510.00    1518.00    4962.00       1.00       1510.00        940.00   

        NASDelay  SecurityDelay  LateAircraftDelay  
count  128859.00      128859.00          128859.00  
mean       14.96           0.09              25.54  
std        33.35           1.83              42.41  
min         0.00           0.00               0.00  
25%         0.00           0.00               0.00  
50%         2.00           0.00               8.00  
75%        15.00           0.00              34.00  
max       621.00         211.00            1143.00  

[8 rows x 21 columns]

In [272]:
flights_df.isnull().sum() ##we count the number of values

Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                703
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
ActualElapsedTime      831
CRSElapsedTime          16
AirTime                831
ArrDelay               831
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
Cancelled                0
CarrierDelay         71141
WeatherDelay         71141
NASDelay             71141
SecurityDelay        71141
LateAircraftDelay    71141
dtype: int64

### Handeling NaN values

There are several options, each of them should be considered carefully, and probably non of them is good enough, but is what we have...

- Remove rows with NaNs and NoNs
- Fill them with another value (-99999, 0)
- Do an interpolation (Mean)


Delays's. Since delay is just registered from an specific time higher than 10-15 min there is a lot of delays. In this case it makes more sence assigne a = for the flights with no delay

In the case of Arrival Time Nans, this are very rare cases, that for some probably) external reasons we don't have them. These spetial cases are not gona give us any information, it would be more like outliers, so it makes more sence to delete them


In [273]:
## I create a dictionary with the columns I want to fillna and its value. And fill them
values = {"CarrierDelay": 0, "WeatherDelay": 0, "NASDelay": 0, "SecurityDelay": 0, "LateAircraftDelay": 0}
flights_df = flights_df.fillna(value=values)

In [274]:
## I drop the rest of the rows with nan's
flights_df = flights_df.dropna()

In [275]:
##Checking of the NaN's of the df
flights_df.isnull().sum() ##we count the number of values

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
Cancelled            0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

In [276]:
## And checking that the DF still correct
flights_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 199169 entries, 0 to 199999
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Year               199169 non-null  int64  
 1   Month              199169 non-null  int64  
 2   DayofMonth         199169 non-null  int64  
 3   DayOfWeek          199169 non-null  int64  
 4   DepTime            199169 non-null  float64
 5   CRSDepTime         199169 non-null  int64  
 6   ArrTime            199169 non-null  float64
 7   CRSArrTime         199169 non-null  int64  
 8   UniqueCarrier      199169 non-null  object 
 9   FlightNum          199169 non-null  int64  
 10  ActualElapsedTime  199169 non-null  float64
 11  CRSElapsedTime     199169 non-null  float64
 12  AirTime            199169 non-null  float64
 13  ArrDelay           199169 non-null  float64
 14  DepDelay           199169 non-null  float64
 15  Origin             199169 non-null  object 
 16  De

**Avarege flight velocity**

For this, I will take the Air time and the distance using the formula Vel = Distance/Time
Distance is in miles
Air Time is in minuts

In [277]:
flights_df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'Cancelled', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [278]:
##I check for errors and see if there is any negative value or 0
flights_df.loc[flights_df["AirTime"]<=0]

Empty DataFrame
Columns: [Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, CRSArrTime, UniqueCarrier, FlightNum, ActualElapsedTime, CRSElapsedTime, AirTime, ArrDelay, DepDelay, Origin, Dest, Distance, Cancelled, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay]
Index: []

[0 rows x 24 columns]

In [279]:
flights_df['AirTime'].describe().round(3)

count    199169.000
mean        108.365
std          68.596
min           1.000
25%          58.000
50%          90.000
75%         137.000
max         654.000
Name: AirTime, dtype: float64

In [280]:
flights_df['Distance'].describe().round(3)

count    199169.000
mean        765.708
std         573.670
min          31.000
25%         341.000
50%         606.000
75%         997.000
max        4962.000
Name: Distance, dtype: float64

In [281]:
## Calcution of the Avarage velocity
flights_df = flights_df.copy() ##In order to avoid the copy of slice of df Warning
##I apply the formula and multiply by 60 in order to have velocities in miles/hour
flights_df['AveFliVelocity'] = (flights_df['Distance']/flights_df['AirTime'])*60

In [282]:
flights_df['AveFliVelocity'].describe().round(3)

count    199169.000
mean        397.106
std          81.659
min          33.176
25%         352.059
50%         403.960
75%         448.333
max        8407.500
Name: AveFliVelocity, dtype: float64

I just have a quicke check on the statistics of the new column Avarage Flight:
   - We see a mean of 397 miles/hour and most of the data up to the 75% is until 450 miles/hour
   - We see there are also some outliers because the minimum and maximum values makes little sence
   

**Was the flight delayed?**

In order to do a new column registering if the flight was or wasnot delayed, I will compute all the different Types of Delays ('CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay') see if any of this happends and put the new column with True if yes False if not.

For now, I will not use ArrDelay and DeptDelay because I assume its already in the columns of the reason of the delay, but when I have the column with all the delays computed I'll check if this assumtion is true

In [283]:
flights_df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'Cancelled', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay',
       'AveFliVelocity'],
      dtype='object')

In [286]:
flights_df.loc[flights_df["CarrierDelay"] != 0]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
5       2008      3          17          1   1513.0        1415   1714.0   
9       2008      6          27          5   1801.0        1745   1928.0   
12      2008      5          28          3   1353.0        1245   1712.0   
13      2008     10           5          7   1646.0        1530   1719.0   
14      2008      4           3          4   1110.0        1100   1256.0   
...      ...    ...         ...        ...      ...         ...      ...   
199989  2008      6          12          4   1932.0        1915   2024.0   
199990  2008      2          24          7   1352.0        1320   1506.0   
199991  2008      5           9          5   1937.0        1820   2146.0   
199992  2008      7          12          6   2023.0        1940   2234.0   
199998  2008     12          15          1   1007.0         937   1116.0   

        CRSArrTime UniqueCarrier  FlightNum  ...  Origin  Dest  Distance  \
5             1605            WN        364  ...     MCI   PDX      1481   
9             1910            WN       2049  ...     SAN   SMF       480   
12            1600            AA          5  ...     DFW   HNL      3784   
13            1615            WN        333  ...     BUF   MDW       468   
14            1220            DL        781  ...     ATL   DEN      1199   
...            ...           ...        ...  ...     ...   ...       ...   
199989        2008            OO       5580  ...     LAX   SAN       109   
199990        1425            WN       2553  ...     PHX   SJC       621   
199991        2034            US       1724  ...     TPA   DCA       814   
199992        2212            XE       2076  ...     EWR   IND       644   
199998        1000            9E       5821  ...     ATL   JAN       341   

        Cancelled  CarrierDelay WeatherDelay NASDelay  SecurityDelay  \
5               0          14.0          0.0     11.0            0.0   
9               0          13.0          0.0      2.0            0.0   
12              0          68.0          0.0      4.0            0.0   
13              0           1.0          0.0      0.0            0.0   
14              0          10.0          0.0     26.0            0.0   
...           ...           ...          ...      ...            ...   
199989          0          16.0          0.0      0.0            0.0   
199990          0          12.0          0.0      9.0            0.0   
199991          0           7.0          0.0      0.0            0.0   
199992          0           5.0          0.0      0.0            0.0   
199998          0          30.0          0.0     46.0            0.0   

        LateAircraftDelay  AveFliVelocity  
5                    44.0      393.185841  
9                     3.0      405.633803  
12                    0.0      473.000000  
13                   63.0      346.666667  
14                    0.0      399.666667  
...                   ...             ...  
199989                0.0      204.375000  
199990               20.0      318.461538  
199991               65.0      440.000000  
199992               17.0      390.303030  
199998                0.0      319.687500  

[67486 rows x 25 columns]

In [287]:
##I create the condition that I want to look for, so the delay columns different from 0 since each of this column is in miunutes of delay
##Then I create the new column with True if any of the delayd columns is diferent from 0 and fals if all of them are 0
condition = (flights_df['CarrierDelay'] != 0) | (flights_df['WeatherDelay'] != 0) | (flights_df['NASDelay'] != 0) | (flights_df['SecurityDelay'] != 0) | (flights_df['LateAircraftDelay'] != 0) 
flights_df['FlightDelayed'] = np.where(condition, True, False)

In [288]:
flights_df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0       2008      4          23          3   1555.0        1515   1658.0   
1       2008     12          21          7   1612.0        1435   1719.0   
2       2008      1          15          2   2250.0        2215     48.0   
3       2008     10           4          6   1015.0        1005   1106.0   
4       2008      6           6          5   1915.0        1725   2040.0   
...      ...    ...         ...        ...      ...         ...      ...   
199995  2008     11          26          3   1114.0        1105   1404.0   
199996  2008      3          22          6   1222.0        1215   1521.0   
199997  2008      2          27          3   1821.0        1803   1935.0   
199998  2008     12          15          1   1007.0         937   1116.0   
199999  2008      2           4          1   2232.0        2225   2358.0   

        CRSArrTime UniqueCarrier  FlightNum  ...  Dest  Distance  Cancelled  \
0             1605            WN        649  ...   AUS       189          0   
1             1559            MQ       4750  ...   JFK       213          0   
2               24            US       1116  ...   MHT       737          0   
3             1100            WN        185  ...   PIT       210          0   
4             1850            AA       1618  ...   MCI       403          0   
...            ...           ...        ...  ...   ...       ...        ...   
199995        1355            WN       3116  ...   PHX       646          0   
199996        1507            UA        490  ...   DFW       641          0   
199997        1953            CO       1600  ...   CLE       363          0   
199998        1000            9E       5821  ...   JAN       341          0   
199999        2357            NW       1469  ...   MOT       449          0   

        CarrierDelay  WeatherDelay NASDelay SecurityDelay  LateAircraftDelay  \
0                0.0           0.0     13.0           0.0               40.0   
1                0.0           0.0     49.0           0.0               31.0   
2                0.0           0.0      0.0           0.0               24.0   
3                0.0           0.0      0.0           0.0                0.0   
4                0.0           0.0     46.0           0.0               64.0   
...              ...           ...      ...           ...                ...   
199995           0.0           0.0      0.0           0.0                0.0   
199996           0.0           0.0      0.0           0.0                0.0   
199997           0.0           0.0      0.0           0.0                0.0   
199998          30.0           0.0     46.0           0.0                0.0   
199999           0.0           0.0      0.0           0.0                0.0   

        AveFliVelocity  FlightDelayed  
0           324.000000           True  
1           311.707317           True  
2           446.666667           True  
3           293.023256          False  
4           366.363636           True  
...                ...            ...  
199995      399.587629          False  
199996      452.470588          False  
199997      403.333333          False  
199998      319.687500           True  
199999      384.857143          False  

[199169 rows x 26 columns]

I want to check if arrival and departure delay is included in the different delay reasions

In [289]:
## Selection of the columns I want to check
flights_df[['ArrDelay', 'DepDelay', 'FlightDelayed']]

ArrDelay  DepDelay  FlightDelayed
0           53.0      40.0           True
1           80.0      97.0           True
2           24.0      35.0           True
3            6.0      10.0          False
4          110.0     110.0           True
...          ...       ...            ...
199995       9.0       9.0          False
199996      14.0       7.0          False
199997     -18.0      18.0          False
199998      76.0      30.0           True
199999       1.0       7.0          False

[199169 rows x 3 columns]

- At first sight we see that the arribal an departure delays are included, since for instance row 3 where the arr and dep delays are les than 15 the flightDelayed column is False.
- But we also see that they compute as no delay that the sum of arrival and departure delay is beloww 15 like in row 199997.This is something to coinsider for future studies on the delay

In [290]:
## “len()”method counts the number of rows. 
##Inside its braces, we specified a condition with the name of the DataFrame and the DataFrame column name.
len(flights_df.loc[flights_df['FlightDelayed']== True])

128859

In [291]:
len(flights_df.loc[flights_df['FlightDelayed']== False])

70310

Checking the amount of row with delay and not delay we see there is more delayed flights than not!!

### A table with the Carriers with more delay

To check this I want to group the data by carrier and sort by the amout of delays

In [292]:
##I groupby Unique Carrier, sum the values so I can have a sum of the True's in Flight Delayed,
##and sort descending by Flight delayed
flights_df.groupby('UniqueCarrier').sum().sort_values(['FlightDelayed'], ascending=False)

Year   Month  DayofMonth  DayOfWeek     DepTime  \
UniqueCarrier                                                        
WN             78372240  238905      618543     157297  60996534.0   
AA             39589728  117270      311891      78170  30037711.0   
MQ             29130056   86501      227874      56214  21315475.0   
UA             29304752   86955      227940      58368  22069067.0   
OO             27198360   82808      213275      54156  20068085.0   
XE             21423352   61153      166840      41189  16244534.0   
DL             23786768   77946      186535      48049  17959687.0   
US             20407304   61825      159989      40070  15150020.0   
CO             20441440   64203      164173      39774  15344386.0   
EV             16901336   52297      132461      33723  12575375.0   
NW             16409376   48748      126703      32115  12261222.0   
YV             13861224   44047      107834      28247  10333863.0   
FL             14979680   46989      119986      30520  11972863.0   
OH             11027936   32724       85686      21391   8382194.0   
B6             11100224   35002       88139      22155   8634021.0   
9E             10539992   30171       80506      20802   7707661.0   
AS              8080192   25546       62965      15971   5997183.0   
F9              5686656   17684       43502      11524   4276569.0   
HA              1546160    6055       12648       3288   1114021.0   
AQ               144576     101         889        320    104797.0   

               CRSDepTime     ArrTime  CRSArrTime  FlightNum  \
UniqueCarrier                                                  
WN               59161815  64189591.0    65097410   63019625   
AA               29008329  31927302.0    32889066   24872438   
MQ               20338294  22983042.0    22634706   58099601   
UA               21230399  23054453.0    23642174   10511657   
OO               19179290  21293696.0    21179908   72458189   
XE               15455909  17374277.0    17510730   27838166   
DL               17411928  19323230.0    19709424   12895943   
US               15113938  16127846.0    16578824   10322355   
CO               14861734  16529865.0    17276586    9385121   
EV               11997427  13494221.0    13319039   39694733   
NW               11871870  13161478.0    13415129    6672128   
YV                9942217  10682957.0    10763997   36573635   
FL               11712810  12373367.0    12665085    3611070   
OH                8020466   8971209.0     8995636   30935814   
B6                8573217   8395508.0     9110074    2871919   
9E                7349223   8296775.0     8217278   22759391   
AS                5829955   6429252.0     6399882    1266307   
F9                4263516   4585453.0     4846703    1405124   
HA                1083630   1200488.0     1233892     165520   
AQ                 102847    109987.0      110838      21254   

               ActualElapsedTime  ...   DepDelay  Distance  Cancelled  \
UniqueCarrier                     ...                                   
WN                     4289617.0  ...  1348679.0  25468491          0   
AA                     3368471.0  ...   914542.0  21246217          0   
MQ                     1390827.0  ...   637421.0   6338539          0   
UA                     2480806.0  ...   722388.0  15733471          0   
OO                     1290271.0  ...   604195.0   6033464          0   
XE                     1266703.0  ...   541145.0   6365665          0   
DL                     1939223.0  ...   465519.0  11464676          0   
US                     1625446.0  ...   389362.0   9902710          0   
CO                     2000339.0  ...   442332.0  12591504          0   
EV                      825426.0  ...   406628.0   3901065          0   
NW                     1195116.0  ...   324832.0   6793560          0   
YV                      628254.0  ...   387446.0   2826592          0   
FL                      966264.0  ...   319893.

In [293]:
##If I just one the key column of the group by and the Flight delay I also can do it like this
CarrierDelayed = flights_df.groupby('UniqueCarrier')['FlightDelayed'].sum()
CarrierDelayed.sort_values(ascending=False)

UniqueCarrier
WN    21100
AA    13659
MQ     9964
UA     9836
OO     9133
XE     7531
DL     7441
US     6173
CO     6023
EV     5866
NW     5684
YV     5239
FL     4994
OH     4109
B6     3885
9E     3604
AS     2532
F9     1618
HA      442
AQ       26
Name: FlightDelayed, dtype: int64

We can See that the Carrier that accumulates more delays is WN and the one with less is AQ

In [294]:
##We can directly use describe() methode to perform a set of aggregations that describe each group in the data
##I creat a new df to be able to sort it
CarrierStats = flights_df.groupby('UniqueCarrier')['FlightDelayed'].describe()
CarrierStats.sort_values('freq', ascending= False)

count unique    top   freq
UniqueCarrier                            
WN             39030      2   True  21100
AA             19716      2   True  13659
MQ             14507      2   True   9964
UA             14594      2   True   9836
OO             13545      2   True   9133
XE             10669      2   True   7531
DL             11846      2   True   7441
US             10163      2   True   6173
CO             10180      2   True   6023
EV              8417      2   True   5866
NW              8172      2   True   5684
YV              6903      2   True   5239
FL              7460      2   True   4994
OH              5492      2   True   4109
B6              5528      2   True   3885
9E              5249      2   True   3604
AS              4024      2   True   2532
F9              2832      2   True   1618
HA               770      2   True    442
AQ                72      2  False     46

Here we see the basics statistic of the discribe for the table. The count referes to the number of times the Carrier had true or False, uniques is that we just have 2 objects (true and False) te top referes to the one of the to that appears more times and the frequency is the number of times it appears.
An interesting thing is that most of the times the frequency and count are both in the descending order, but not allways!
So normally if they fly more they have more delays, but some times, they fly less and have more delays like like US inrespect to CO.

### Which are the longest flights? The ones with more delay? Look for the longuest routes and the ones that have more delays.
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [296]:
flights_df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'Cancelled', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay',
       'AveFliVelocity', 'FlightDelayed'],
      dtype='object')

#### Which is the longest flight?

In [302]:
##Here I show the 10 longest flights by flight number
flights_df.groupby('FlightNum').sum().sort_values('Distance', ascending=False)[:10]

Year  Month  DayofMonth  DayOfWeek   DepTime  CRSDepTime  \
FlightNum                                                               
15         327304   1025        2260        632  244420.0      235652   
73         253008    756        1971        561  162492.0      152728   
16         321280   1006        2564        636  216425.0      207386   
3          220880    708        1627        437  138291.0      130799   
1          190760    589        1474        374   98292.0       91849   
29         311240    977        2401        590  252402.0      250670   
47         299192   1012        2606        612  264819.0      258884   
40         327304    994        2650        649  258415.0      248206   
75         303208    921        2418        630  244715.0      233746   
14         190760    590        1390        390  156216.0      151627   

            ArrTime  CRSArrTime  ActualElapsedTime  CRSElapsedTime  ...  \
FlightNum                                                           ...   
15         290046.0      286144            45166.0         45232.0  ...   
73         189747.0      186916            36595.0         36688.0  ...   
16         260742.0      263473            35557.0         35927.0  ...   
3          162844.0      155868            27993.0         27989.0  ...   
1          128713.0      122702            27681.0         28073.0  ...   
29         249068.0      285841            29806.0         30631.0  ...   
47         261570.0      282652            28291.0         28788.0  ...   
40         291929.0      294603            27174.0         27617.0  ...   
75         263345.0      256222            27950.0         27956.0  ...   
14         130207.0      135957            23151.0         23490.0  ...   

           DepDelay  Distance  Cancelled  CarrierDelay  WeatherDelay  \
FlightNum                                                              
15           5528.0    312152          0        1033.0         214.0   
73           6084.0    256460          0        3117.0          61.0   
16           5639.0    253014          0        1968.0         156.0   
3            4452.0    199987          0        1428.0         134.0   
1            4003.0    198745          0        1768.0         225.0   
29           9372.0    184646          0        1212.0        1052.0   
47           6855.0    182890          0        1678.0         264.0   
40           6209.0    178712          0        1265.0          10.0   
75           7209.0    174138          0        1320.0         137.0   
14           4229.0    174098          0         441.0           0.0   

           NASDelay  SecurityDelay  LateAircraftDelay  AveFliVelocity  \
FlightNum                                                               
15           1355.0            0.0             2823.0    70909.086985   
73            833.0            8.0             1761.0    54316.403593   
16           1177.0            0.0             1672.0    72549.763479   
3            1542.0           13.0             1318.0    47656.829120   
1            1318.0           12.0              344.0    42542.254980   
29           1695.0            0.0             4706.0    64526.318817   
47            687.0           24.0             3553.0    64622.494340   
40           1106.0           38.0             3165.0    68164.554057   
75           2298.0            0.0             3295.0    60463.849706   
14           1931.0            0.0             1519.0    44439.822888   

           FlightDelayed  
FlightNum                 
15                   104  
73                    73  
16                    89  
3                     77  
1                     51  
29                    97  
47                    93  
40                    87  
75                   107  
14                    53  

[10 rows x 22 columns]

#### Which is the flight with more delay?

In [303]:
##Here I show the 10 flights with more delay by flight number
flights_df.groupby('FlightNum').sum().sort_values('FlightDelayed', ascending=False)[:10]

Year  Month  DayofMonth  DayOfWeek   DepTime  CRSDepTime  \
FlightNum                                                               
50         347384   1051        2674        698  288245.0      279449   
75         303208    921        2418        630  244715.0      233746   
15         327304   1025        2260        632  244420.0      235652   
44         281120    940        2188        570  220568.0      214978   
403        261040    831        2035        508  182488.0      177388   
29         311240    977        2401        590  252402.0      250670   
511        271080    788        2039        552  220575.0      213987   
47         299192   1012        2606        612  264819.0      258884   
321        265056    796        2063        556  226797.0      220553   
701        271080    814        2318        493  200066.0      197401   

            ArrTime  CRSArrTime  ActualElapsedTime  CRSElapsedTime  ...  \
FlightNum                                                           ...   
50         328434.0      326778            25529.0         25589.0  ...   
75         263345.0      256222            27950.0         27956.0  ...   
15         290046.0      286144            45166.0         45232.0  ...   
44         243488.0      237456            22280.0         22398.0  ...   
403        175949.0      196027            19341.0         19159.0  ...   
29         249068.0      285841            29806.0         30631.0  ...   
511        234008.0      241265            21033.0         21118.0  ...   
47         261570.0      282652            28291.0         28788.0  ...   
321        239315.0      252431            22858.0         23058.0  ...   
701        225444.0      217599            21878.0         22058.0  ...   

           DepDelay  Distance  Cancelled  CarrierDelay  WeatherDelay  \
FlightNum                                                              
50           6996.0    156827          0        1030.0          12.0   
75           7209.0    174138          0        1320.0         137.0   
15           5528.0    312152          0        1033.0         214.0   
44           6550.0    152026          0        1434.0         504.0   
403          4860.0    111580          0        1844.0          60.0   
29           9372.0    184646          0        1212.0        1052.0   
511          5468.0    119736          0        1458.0         209.0   
47           6855.0    182890          0        1678.0         264.0   
321          6924.0    138316          0        1292.0         313.0   
701          4705.0    132547          0         845.0          10.0   

           NASDelay  SecurityDelay  LateAircraftDelay  AveFliVelocity  \
FlightNum                                                               
50           2044.0            0.0             3658.0    71602.800654   
75           2298.0            0.0             3295.0    60463.849706   
15           1355.0            0.0             2823.0    70909.086985   
44           1082.0            0.0             3374.0    61155.317964   
403          1250.0            0.0             1806.0    53162.134658   
29           1695.0            0.0             4706.0    64526.318817   
511          1038.0            0.0             2637.0    54669.636770   
47            687.0           24.0             3553.0    64622.494340   
321          1424.0            0.0             3615.0    54024.335014   
701          1232.0            1.0             2360.0    54284.694108   

           FlightDelayed  
FlightNum                 
50                   125  
75                   107  
15                   104  
44                   100  
403                  100  
29                    97  
511                   96  
47                    93  
321                   93  
701                   92  

[10 rows x 22 columns]

#### Which is the longest route?

In [259]:
##Here I groupby two columns Origine and Destination, because I want to check which are the longest routs
##I do it in 2 steps, to see what is doing, first grouping by and aggregating by sum
flights_df.groupby(['Origin', 'Dest']).sum()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
Origin Dest                                                             
ABE    ATL   56224    181         468        115  43554.0       41034   
       CLE   14056     35          74         23  10726.0       10358   
       CLT    2008      6          27          5   1150.0         850   
       CVG    2008      5           2          5   1736.0        1710   
       DTW   26104     71         216         47  20123.0       18997   
...            ...    ...         ...        ...      ...         ...   
YUM    IPL   10040     16          83          9   7257.0        7011   
       LAS    2008      5          30          5   2200.0        2143   
       LAX   26104     83         193         49  16373.0       15376   
       PHX   66264    195         432        133  42873.0       40517   
       SLC   22088     50         114         44  12709.0       12285   

             ArrTime  CRSArrTime  ActualElapsedTime  CRSElapsedTime  ...  \
Origin Dest                                                          ...   
ABE    ATL   50383.0       47176             4069.0          3782.0  ...   
       CLE   11696.0       11207              570.0           529.0  ...   
       CLT    1314.0        1023               84.0            93.0  ...   
       CVG    1920.0        1855              104.0           105.0  ...   
       DTW   22807.0       21295             1564.0          1418.0  ...   
...              ...         ...                ...             ...  ...   
YUM    IPL    7379.0        7096              162.0           165.0  ...   
       LAS    2315.0        2304               75.0            81.0  ...   
       LAX   17305.0       16320              992.0          1004.0  ...   
       PHX   46368.0       43895             2055.0          2058.0  ...   
       SLC   15226.0       14862             1097.0          1197.0  ...   

             DepDelay  Distance  Cancelled  CarrierDelay  WeatherDelay  \
Origin Dest                                                              
ABE    ATL     1640.0     19376          0        1046.0         144.0   
       CLE      208.0      2373          0         106.0           0.0   
       CLT      180.0       481          0         171.0           0.0   
       CVG       26.0       503          0          25.0           0.0   
       DTW      766.0      5512          0         188.0           0.0   
...               ...       ...        ...           ...           ...   
YUM    IPL      126.0       290          0         111.0           0.0   
       LAS       17.0       238          0           0.0           0.0   
       LAX      597.0      3081          0         268.0           0.0   
       PHX     1636.0      5280          0        1263.0           0.0   
       SLC      264.0      6380          0          37.0           0.0   

             NASDelay  SecurityDelay  LateAircraftDelay  AveFliVelocity  \
Origin Dest                                                               
ABE    ATL      406.0            0.0              304.0    10482.169910   
       CLE       53.0            0.0               85.0     2296.910389   
       CLT        0.0            0.0                0.0      444.000000   
       CVG        0.0            0.0                0.0      355.058824   
       DTW      547.0            0.0              159.0     4193.783813   
...               ...            ...                ...             ...   
YUM    IPL        0.0            0.0                0.0     1040.872549   
       LAS        0.0            0.0                0.0      264.444444   
       LAX      153.0           29.0               87.0     3208.356468   
       PHX      279.0            0.0                0.0     8551.385734   
       SLC        0.0           20.0              116.0     4660.783640   

             FlightDelayed  
Origin Dest                 
ABE    ATL              22  
       CLE               5  
       CLT          

In [263]:
##Second step is to sort by the distance column and I put the table in a new df
RouteDistance = flights_df.groupby(['Origin', 'Dest']).sum().sort_values('Distance', ascending=False)
##Here I just show the 3 key columns and the distance on the new df, and just the first 10 rows
RouteDistance[['Distance']][:10]

Distance
Origin Dest          
LAX    JFK     616275
ORD    LAX     612495
JFK    LAX     594000
SFO    JFK     553404
LAX    ORD     499070
ORD    SFO     496574
LAX    HNL     449856
ATL    LAX     439796
JFK    SFO     431862
SFO    ORD     413504

In the previous table we see the routes with the longest distance

#### Which is the route with more delay?

In [305]:
##Here I groupby two columns Origine and Destination, and sort by the flight delay column and I put the table in a new df
RouteDistance = flights_df.groupby(['Origin', 'Dest']).sum().sort_values('FlightDelayed', ascending=False)
##Here I just show the 2 key columns and the Flight delayed on the new df, and just the first 10 rows
RouteDistance[['FlightDelayed']][:10]

FlightDelayed
Origin Dest               
ORD    LGA             352
LAX    SFO             340
SFO    LAX             296
ATL    LGA             287
ORD    EWR             282
LGA    ORD             279
ATL    EWR             279
DFW    ORD             250
ORD    LAX             231
LAS    LAX             230

In the previous table we see the routes whci acumulate more delays 

## Exercici 3

<div class="alert alert-block alert-info">Exporta el dataset net i amb les noves columnes a Excel.   
</div>
    
    


In [306]:
flights_df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0       2008      4          23          3   1555.0        1515   1658.0   
1       2008     12          21          7   1612.0        1435   1719.0   
2       2008      1          15          2   2250.0        2215     48.0   
3       2008     10           4          6   1015.0        1005   1106.0   
4       2008      6           6          5   1915.0        1725   2040.0   
...      ...    ...         ...        ...      ...         ...      ...   
199995  2008     11          26          3   1114.0        1105   1404.0   
199996  2008      3          22          6   1222.0        1215   1521.0   
199997  2008      2          27          3   1821.0        1803   1935.0   
199998  2008     12          15          1   1007.0         937   1116.0   
199999  2008      2           4          1   2232.0        2225   2358.0   

        CRSArrTime UniqueCarrier  FlightNum  ...  Dest  Distance  Cancelled  \
0             1605            WN        649  ...   AUS       189          0   
1             1559            MQ       4750  ...   JFK       213          0   
2               24            US       1116  ...   MHT       737          0   
3             1100            WN        185  ...   PIT       210          0   
4             1850            AA       1618  ...   MCI       403          0   
...            ...           ...        ...  ...   ...       ...        ...   
199995        1355            WN       3116  ...   PHX       646          0   
199996        1507            UA        490  ...   DFW       641          0   
199997        1953            CO       1600  ...   CLE       363          0   
199998        1000            9E       5821  ...   JAN       341          0   
199999        2357            NW       1469  ...   MOT       449          0   

        CarrierDelay  WeatherDelay NASDelay SecurityDelay  LateAircraftDelay  \
0                0.0           0.0     13.0           0.0               40.0   
1                0.0           0.0     49.0           0.0               31.0   
2                0.0           0.0      0.0           0.0               24.0   
3                0.0           0.0      0.0           0.0                0.0   
4                0.0           0.0     46.0           0.0               64.0   
...              ...           ...      ...           ...                ...   
199995           0.0           0.0      0.0           0.0                0.0   
199996           0.0           0.0      0.0           0.0                0.0   
199997           0.0           0.0      0.0           0.0                0.0   
199998          30.0           0.0     46.0           0.0                0.0   
199999           0.0           0.0      0.0           0.0                0.0   

        AveFliVelocity  FlightDelayed  
0           324.000000           True  
1           311.707317           True  
2           446.666667           True  
3           293.023256          False  
4           366.363636           True  
...                ...            ...  
199995      399.587629          False  
199996      452.470588          False  
199997      403.333333          False  
199998      319.687500           True  
199999      384.857143          False  

[199169 rows x 26 columns]

In [308]:
flights_df.to_excel('FlightSummary.xlsx', sheet_name = 'Sheet1')